src/index_runner/event_loop.py

```diff
@@ -0,0 +1,124 @@
+'''
+The event loop for the index runner.
+'''
+
+import json
+import logging
+import os
+import requests
+import time
+import traceback
+
+from logging import Logger
+from typing import Callable, Dict, Any
+
+from confluent_kafka import Consumer, KafkaError
+
+from src.utils.config import config
+from src.utils.service_utils import wait_for_dependencies
+
+Message = Dict[str, Any]
+
+# TODO TEST unit tests
+
+
+def start_loop(
+        consumer: Consumer,
+        message_handler: Callable[[Message], None],
+        on_success: Callable[[Message], None] = lambda msg: None,
+        on_failure: Callable[[Message, Exception], None] = lambda msg, e: None,
+        on_ready: Callable[[], None] = lambda: None,
+        on_config_update: Callable[[], None] = lambda: None,
+        logger: Logger = logging.getLogger('src.event_loop')):
+    """
+    Run the indexer event loop.
+
+    :param consumer: A Kafka consumer which will be polled for messages.
+    :param message_handler: a processor for messages from Kafka.
+    :param on_success: called after message_handler has returned sucessfully and the message
+        offset has been committed to Kafka. A noop by default.
+    :param on_failure: called if the message_handler, the Kafka commit, or on_success throws an
+        exception. A noop by default.
+    :param on_ready: called when the event loop is intialized and about to start.
+    :param on_config_update: called when the configuration has been updated.
+    :param logger: a logger to use for logging events. By default a standard logger for
+        'src.event_loop'.
+    """
+    # Remove the ready indicator file if it has been written on a previous boot
+    if os.path.exists(config()['proc_ready_path']):
+        os.remove(config()['proc_ready_path'])
+    # Wait for dependency services (ES and RE) to be live
+    wait_for_dependencies(timeout=180)
+    # Used for re-fetching the configuration with a throttle
+    last_updated_minute = int(time.time() / 60)
+    if not config()['global_config_url']:
+        config_tag = _fetch_latest_config_tag()
+    on_ready()
+    # Touch a temp file indicating the daemon is ready
+    with open(config()['proc_ready_path'], 'w') as fd:
+        fd.write('')
+
```

Comment:
I think I should move everything above this line into main.py other than the `config_tag` stuff. That would make the loop a lot easier to test - I'd just have to mock requests in a couple places.

The other thing I'm thinking is make config() a parameter as well so that can be mocked, but I'll do that in a later commit if so, and also hopefully after we figure out what the plan is re config handling as that might mean changes here.

Reply:
Definitely makes sense to move wait_for_dependencies() out into main.py

It makes sense to me that the config init, fetching, and re-fetching code lives here

Down the line, we could make the `config()` function itself check a timestamp of last refresh, and if it's expired do a re-fetch. So there would not have to be any of that logic in this loop. Not sure why we didn't do that in the first place.

Reply:
moved the proc_ready, dependencies, and on_ready stuff to main.py